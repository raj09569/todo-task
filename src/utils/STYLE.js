/**
 * This file contains
 * reusable styles
 */
module.exports = {
  mLeft10: {
    marginLeft: 10,
  },
};
