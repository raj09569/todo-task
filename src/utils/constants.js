/**
 * This file contains
 * all application constants
 */
module.exports = {
  /**
   * All routes
   * will be defined here
   */
  ROUTE_NAMES: {
    Splash: 'Splash',
    List: 'List',
    Add: 'Add',
  },
  /**
   * All app colors
   * will be defined here
   */
  COLORS: {
    primaryColor: '#03A9F4',
    tranperant: 'transperant',
    formGroup: '#f2f2f2',
    formLabel: '#708090',
    inputBg: '#e6f7ff',
    listItemBgColor: '#F5F5F5',
    listItemBorderColor: '#999999',
    activeStatus: '#28A745',
    pendingStatus: '#FFC107',
    descTxtColor: '#333333',
    black: '#000000',
    white: '#ffffff',
    btnBorderColor: '#0056b3',
    btnBgColor: '#007bff',
    errorColor: '#ff0000',
    formBgColor: '#cccccc',
  },
  /**
   * All app fontsizes
   * will be defined here
   */
  FONT_SIZES: {
    xsmall: 10,
    small: 12,
    normal: 14,
    large: 18,
    medium: 16,
    xLarge: 20,
    xxLarge: 22,
  },
  /**
   * All app session keys
   * will be defined here
   */
  SESSION_KEYS: {
    todoList: 'todoList',
  },
};
