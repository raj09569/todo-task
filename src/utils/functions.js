import {useNavigationContainerRef} from '@react-navigation/native';
import * as LocalAuthentication from 'expo-local-authentication';
import {Platform} from 'react-native';

/**
 *
 * @param {*} length
 * @returns
 * random string
 * which generated in defined length
 */
const randomString = length => {
  const characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let result = '';
  for (let i = 0; i < length; i++) {
    const randomIndex = Math.floor(Math.random() * characters.length);
    result += characters.charAt(randomIndex);
  }
  return result;
};

/**
 *
 * @returns
 * its a common function
 * to perform local device authentication
 * in the app.
 * it returns response or error
 */
const validateAuth = async () => {
  /**
   * check device compatability
   */
  const deviceCompatibility = await LocalAuthentication.hasHardwareAsync();
  if (deviceCompatibility) {
    try {
      /**
       * Local device authentication
       * based upon device platform
       */
      if (Platform.OS === 'ios') {
        let result = await LocalAuthentication.authenticateAsync({
          promptMessage: 'Please authenticate here',
        });
        return result;
      } else {
        let result = await LocalAuthentication.authenticateAsync({
          cancelLabel: 'Cancel',
          promptMessage: 'Please authenticate here',
        });
        return result;
      }
    } catch (err) {
      return {
        error: err,
        success: false,
      };
    }
  } else {
    return {
      error: 'device_not_compatible',
      success: false,
    };
  }
};

/**
 * using this common function
 * we will reset stack and
 * navigate to new screen
 */
function resetPush(navigation, routeName) {
  navigation.reset({
    index: 0,
    routes: [{name: routeName}],
  });
}

export {randomString, validateAuth, resetPush};
