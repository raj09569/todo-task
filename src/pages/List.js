import {useEffect, useState} from 'react';
import {
  Alert,
  FlatList,
  Image,
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import AppHeader from '../components/AppHeader';
import {
  COLORS,
  FONT_SIZES,
  ROUTE_NAMES,
  SESSION_KEYS,
} from '../utils/constants';
import {useNavigation} from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {validateAuth} from '../utils/functions';
import {mLeft10} from '../utils/STYLE';

export default function List() {
  const [todoList, setTodoList] = useState([]);
  const navigation = useNavigation();

  useEffect(() => {
    /**
     * Fetching list of todos
     * from async storage
     */
    fetchTodoList();
  }, []);

  async function fetchTodoList() {
    await AsyncStorage.getItem(SESSION_KEYS.todoList).then(list => {
      if (list) {
        list = JSON.parse(list);
        setTodoList(list);
      }
    });
  }

  function onRefersh() {
    fetchTodoList();
  }

  function deleteTodo(id) {
    /**
     * validating local device authentication
     * to delete todo
     */
    validatingAuth(async () => {
      let newArray = [...todoList];
      newArray.splice(
        newArray.findIndex(todo => (todo.id = id)),
        1,
      );
      await AsyncStorage.setItem(
        SESSION_KEYS.todoList,
        JSON.stringify(newArray),
      ).then(() => {
        fetchTodoList();
      });
    });
  }

  function updateStatus(id) {
    /**
     * validating local device authentication
     * to update todo status
     */
    validatingAuth(async () => {
      let newArray = [...todoList];
      let todo = newArray.find(todo => todo.id == id);
      todo.status = !todo.status;
      await AsyncStorage.setItem(
        SESSION_KEYS.todoList,
        JSON.stringify(newArray),
      ).then(() => {
        fetchTodoList();
      });
    });
  }

  async function validatingAuth(action) {
    const authValidation = await validateAuth();
    if (authValidation.success) {
      /**
       * if authentication got succeded
       * we will perform action
       * which recived as argument
       */
      action();
    } else {
      /**
       * error handling
       */
      if (authValidation.error === 'system_cancel') {
        return Alert.alert(
          'System Cancellation',
          'Authentication was cancelled by System',
        );
      }
      if (authValidation.error === 'user_cancel') {
        return Alert.alert(
          'User Cancellation',
          'Authentication was cancelled by User',
        );
      }
      if (authValidation.error === 'device_not_compatible') {
        return Alert.alert(
          'Device Compatibility',
          'Your device is not compatible',
        );
      }
      return Alert.alert('Error', authValidation.error);
    }
  }

  return (
    <SafeAreaView>
      <AppHeader
        title={'Todo List'}
        showAdd
        showRefresh
        onRefersh={onRefersh}
      />
      <FlatList
        data={todoList}
        keyExtractor={item => item.id}
        style={styles.todoList}
        renderItem={({item}) => {
          const {id, name, description, status} = item;
          return (
            <View style={styles.listItem}>
              <View style={mLeft10}>
                <View style={styles.listRow}>
                  <Text style={styles.name}>{name}</Text>
                  <Text style={styles.id}>{`Task #${id}`}</Text>
                </View>
                <Text style={styles.description}>{description}</Text>
              </View>
              <View
                style={[
                  styles.statusBg,
                  {
                    backgroundColor: status
                      ? COLORS.activeStatus
                      : COLORS.pendingStatus,
                  },
                ]}
              />
              <View style={styles.listButtons}>
                <TouchableOpacity
                  onPress={() =>
                    navigation.push(ROUTE_NAMES.Add, {todoItem: item})
                  }>
                  <Image
                    style={styles.listIcon}
                    source={require('../assets/images/icons/edit.png')}
                  />
                </TouchableOpacity>
                <TouchableOpacity
                  onPress={() => {
                    Alert.alert(
                      'Delete Confirmation',
                      'Would you like to delete todo?',
                      [
                        {
                          text: 'No',
                          style: 'cancel',
                        },
                        {text: 'Yes', onPress: () => deleteTodo(id)},
                      ],
                    );
                  }}>
                  <Image
                    style={styles.listIcon}
                    source={require('../assets/images/icons/delete.png')}
                  />
                </TouchableOpacity>
                <TouchableOpacity
                  onPress={() => {
                    Alert.alert(
                      'Update Status',
                      'Would you like to change todo status?',
                      [
                        {
                          text: 'No',
                          style: 'cancel',
                        },
                        {text: 'Yes', onPress: () => updateStatus(id)},
                      ],
                    );
                  }}>
                  <Image
                    style={styles.listIcon}
                    source={
                      status
                        ? require('../assets/images/icons/checked.png')
                        : require('../assets/images/icons/unchecked.png')
                    }
                  />
                </TouchableOpacity>
              </View>
            </View>
          );
        }}
      />
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  listItem: {
    backgroundColor: COLORS.listItemBgColor,
    borderColor: COLORS.listItemBorderColor,
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    marginTop: 10,
  },
  listIcon: {width: 24, height: 24, marginHorizontal: 5},
  statusBg: {
    width: 10,
    position: 'absolute',
    left: 0,
    bottom: 0,
    top: 0,
    borderTopLeftRadius: 10,
    borderBottomLeftRadius: 10,
  },
  description: {color: COLORS.descTxtColor, fontSize: FONT_SIZES.small},
  id: {
    color: COLORS.black,
    fontSize: FONT_SIZES.small,
    fontWeight: 'bold',
  },
  name: {
    color: COLORS.black,
    fontSize: FONT_SIZES.normal,
    fontWeight: '400',
  },
  listRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  todoList: {
    marginHorizontal: 15,
    marginTop: '5%',
    height: '100%',
  },
  listButtons: {flexDirection: 'row', alignSelf: 'flex-end'},
});
