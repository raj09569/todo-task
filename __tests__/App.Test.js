import React from 'react';

import List from '../src/pages/List';
import {fireEvent, render} from '@testing-library/react-native';
import {ROUTE_NAMES} from '../src/utils/constants';

const mockedNavigate = jest.fn();
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('Listing page ', () => {
  it('it should go to add todo page', () => {
    const page = render(<List />);
    const addTodoButton = page.getByTestId('addTodoButton');
    fireEvent.press(addTodoButton);
    expect(mockedNavigate).toHaveBeenCalledWith(ROUTE_NAMES.Add);
  });
});
