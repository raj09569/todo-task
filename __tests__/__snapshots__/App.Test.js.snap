// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          onGestureCancel={[Function]}
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          {
                            "alignItems": "center",
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <Text
                          style={
                            {
                              "fontSize": 24,
                              "fontWeight": "bold",
                            }
                          }
                        >
                          To Do
                        </Text>
                        <Text
                          style={
                            {
                              "fontSize": 14,
                              "fontWeight": "normal",
                              "marginVertical": 10,
                            }
                          }
                        >
                          V 1.0.0
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
